### PR TITLE
(frontend) add a level 0 factory builder

### PR DESCRIPTION
**Pull request by rlecellier**
## Purpose

We want to upgrade faker and for that, we need to get ride of helix.

## Proposal

replace helix by a minimal factory tool.

note:
* [x] check if these change improve test speed for `js/pages/DashboardCourses/index.spec.tsx`
```shell
✓ renders multiple pages of enrollments and orders (11010 ms)
```
> It change nothing about DashboardCourses test duration.

# Ideas

* Allow factory definition to use generated children properties
* Allow factory override to use generated children properties
* Allow factory override to use `generatedIndex`

### CODE REVIEW

**Review by jbpenrath — COMMENTED** [review]
So good to pull helix out, upgrade faker and solve our uniqueness issues. 🤩

File: src/frontend/js/utils/CoursesHelper/index.spec.ts [hunk 1]
```diff
@@ -1,8 +1,13 @@
 import { CoursesHelper } from 'utils/CoursesHelper/index';
 import { Order } from 'types/Joanie';
 import { EnrollmentFactory, OrderFactory, TargetCourseFactory } from 'utils/test/factories/joanie';
+import { FactoryConfig } from 'utils/test/factories/factories';
 
 describe('CourseHelper', () => {
+  afterEach(() => {
+    FactoryConfig.resetUniqueStore();
+  });
```
[the remark this answers is not in the document]
Comment (rlecellier) [reply src/frontend/js/utils/CoursesHelper/index.spec.ts:9]: I may look at it in a other PR as it'll add lot's of changes.
here the [doc](https://jestjs.io/docs/configuration#setupfilesafterenv-array) for memo

File: src/frontend/js/utils/test/factories/factories.ts [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    : K[attr];
+};
+
+interface FactoryBuilderOptions {
+  generateIndex?: number;
+}
+type FactoryBuilder<TData> = (options?: FactoryBuilderOptions) => TData;
+interface FactoryInterface<TData> {
+  generate(amount?: undefined): any;
+  generate(amount: number): any;
```
[the remark this answers is not in the document]
Comment (rlecellier) [reply src/frontend/js/utils/test/factories/factories.ts:23]: I choose to let them in as a second way to use factories.
It didn't change the fact that we choose to use one() and many() in richie.

You prefere that i save that somewhere else and remove it from richie ?

File: src/frontend/js/utils/test/factories/factories.ts [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    },
+  });
+};
+
+export class FactoryConfig {
+  private static instance: FactoryConfig;
+  static GLOBAL_UNIQUE_STORE: Record<RecordKey, RecordKey> = {};
+
+  // eslint-disable-next-line @typescript-eslint/no-empty-function
+  private constructor() {}
```
[the remark this answers is not in the document]
Comment (rlecellier) [reply src/frontend/js/utils/test/factories/factories.ts]: I followed [mozilla documentation](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Private_class_fields) and replaced it by: 
```js
constructor() {
    if (!FactoryConfig.#isInternalConstructing) {
      throw new TypeError('PrivateConstructor is not constructable');
    }
    FactoryConfig.#isInternalConstructing = false;
  }

  static getInstance(): FactoryConfig {
    if (!FactoryConfig.#instance) {
      FactoryConfig.#isInternalConstructing = true;
      FactoryConfig.#instance = new FactoryConfig();
    }

    return FactoryConfig.#instance;
  }
```

File: src/frontend/js/utils/test/factories/joanie.ts [hunk 4]
```diff
@@ -1,200 +1,305 @@
...
+    maxRetries: 20,
+    store: FactoryConfig.GLOBAL_UNIQUE_STORE,
+  });
+  const courseRuns: CourseRun[] = CourseRunFactory().many(3);
+
+  return {
+    id: faker.helpers.unique(faker.datatype.uuid),
+    code,
+    organizations: [],
+    title: faker.helpers.unique(() => `${faker.random.words(3)}(${code})`),
```
[the remark this answers is not in the document]
Comment (rlecellier) [reply src/frontend/js/utils/test/factories/joanie.ts:55]: I missed this one.

File: src/frontend/js/utils/test/factories/joanie.ts [hunk 5]
```diff
@@ -1,200 +1,305 @@
...
+    code: faker.random.alphaNumeric(5),
+    title: faker.random.words(1),
+  };
+});
+
+export const CertificationDefinitionFactory = factory<CertificateDefinition>(() => {
+  return {
+    id: faker.helpers.unique(faker.datatype.number, undefined, {
+      store: FactoryConfig.GLOBAL_UNIQUE_STORE,
+    }),
```
[the remark this answers is not in the document]
Comment (rlecellier) [reply src/frontend/js/utils/test/factories/joanie.ts:90]: in `types/Joanie.ts`
```js
export interface CertificateDefinition {
  id: number;
  title: string;
  description: string;
}
```

File: src/frontend/js/utils/test/factories/factories.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (jbpenrath) [reply src/frontend/js/utils/test/factories/factories.ts:23]: Yep please, from my point of view we are introducing dead code and it does not help new comers to understand how factory should be used.

File: src/frontend/js/utils/test/factories/factories.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (jbpenrath) [reply src/frontend/js/utils/test/factories/factories.ts]: Just a question. Instead to put all this stuff, why not just use a global const containing the store instance ?

File: src/frontend/js/utils/test/factories/joanie.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (jbpenrath) [reply src/frontend/js/utils/test/factories/joanie.ts:90]: Type is outdated 😅

File: src/frontend/js/components/CourseGlimpse/index.spec.tsx [hunk 6]
```diff
@@ -106,7 +109,7 @@ describe('widgets/Search/components/CourseGlimpse', () => {
 
     // Make sure the component renders and shows the state
     screen.getByRole('heading', { name: 'Course 42', level: 3 });
-    screen.getByText('Archived');
+    screen.getByText(/Archived/);
```
[the remark this answers is not in the document]
Comment (jbpenrath) [reply src/frontend/js/components/CourseGlimpse/index.spec.tsx:107]: 🤔 So we should try to get the exact string

File: src/frontend/js/utils/test/factories/joanie.ts [hunk 7]
```diff
@@ -1,200 +1,305 @@
...
+});
+
+export const CertificateProductFactory = factory<Product>(() => {
+  return {
+    id: faker.helpers.unique(faker.datatype.uuid, undefined, {
+      store: FactoryConfig.GLOBAL_UNIQUE_STORE,
+    }),
+    title: faker.helpers.unique(faker.random.words, [10]),
+    course: faker.helpers.unique(faker.random.alphaNumeric, [5]),
+    type: ProductType.CERTIFICATE,
```
[the remark this answers is not in the document]
Comment (rlecellier) [reply src/frontend/js/utils/test/factories/joanie.ts:116]: It was `ProductType.CERTIFICATE`, this PR should not change this.

[the remark this answers is not in the document]
Comment (jbpenrath) [reply src/frontend/js/utils/test/factories/joanie.ts:109]: Agree as discussed

File: src/frontend/js/components/CourseGlimpse/index.spec.tsx [hunk 8]
```diff
@@ -106,7 +104,13 @@ describe('widgets/Search/components/CourseGlimpse', () => {
 
     // Make sure the component renders and shows the state
     screen.getByRole('heading', { name: 'Course 42', level: 3 });
-    screen.getByText('Archived');
+    const intl = createIntl({ locale: 'en' });
+    const formatedDatetime = intl.formatDate(new Date(course.state.datetime!), {
+      year: 'numeric',
+      month: 'short',
+      day: 'numeric',
+    });
```
Comment (jbpenrath) [inline src/frontend/js/components/CourseGlimpse/index.spec.tsx:112]: In other test we are using the standard API. e.g:
```ts
const dateFormatter = Intl.DateTimeFormat('en', {
    day: '2-digit',
    month: 'short',
    year: 'numeric',
  });
const formatedDatetime = dateFormatter.format(new Date(course.state.datetime!))
```

File: src/frontend/js/components/CourseGlimpseList/index.stories.tsx [hunk 9]
```diff
@@ -1,6 +1,6 @@
 import { Meta, StoryObj } from '@storybook/react';
 import { RichieContextFactory, CourseLightFactory } from 'utils/test/factories/richie';
-import { CourseGlimpseList } from '.';
+import { CourseGlimpseList, getCourseGlimpsListProps } from '.';
```
Comment (jbpenrath) [inline src/frontend/js/components/CourseGlimpseList/index.stories.tsx:3]:
```suggestion
import { CourseGlimpseList, getCourseGlimpseListProps } from '.';
```

File: src/frontend/js/components/CourseGlimpseList/index.stories.tsx [hunk 10]
```diff
@@ -10,8 +10,8 @@ type Story = StoryObj<typeof CourseGlimpseList>;
 
 export const RichieCourseList: Story = {
   args: {
-    context: RichieContextFactory().generate(),
-    courses: CourseLightFactory.generate(10),
+    context: RichieContextFactory().one(),
+    courses: getCourseGlimpsListProps(CourseLightFactory().many(10)),
```
Comment (jbpenrath) [inline src/frontend/js/components/CourseGlimpseList/index.stories.tsx:14]:
```suggestion
    courses: getCourseGlimpseListProps(CourseLightFactory().many(10)),
```

**Pull request merged**